### PR TITLE
dispatcher tests

### PR DESCRIPTION
**Pull request by jacquese**
#### Description
Adds basic tests to the dispatcher, and changes the API to make it slightly nicer to test

#### Release note
probably unnecessary

#### Tests
yes

#### Documentation
sparse comments

#### Primary reviewers
@jessicafalk @samiwh

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+namespace
+{
+constexpr Worker_ComponentId COMPONENT_ID = 1000;
+constexpr Worker_ComponentId OTHER_COMPONENT_ID = 1001;
+constexpr Worker_EntityId ENTITY_ID = 1;
+constexpr Worker_EntityId OTHER_ENTITY_ID = 2;
+constexpr double COMPONENT_VALUE = 3;
+constexpr double OTHER_COMPONENT_VALUE = 4;
+
+TUniquePtr<SpatialGDK::ComponentChange> CreateComponentAddedChange(const Worker_ComponentId ComponentId, const double Value)
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:24]: Why a unique pointer? This is a copyable object that doesn't own anything

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:24]: it was because the data inside is non-copyable, but I think what's actually happening is the pointer is copyable, so it's ok. Changed to be just the `ComponentChange` type rather than a unique pointer.

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	TestFalse("Callback was not invoked again", Invoked);
+
+	return true;
+}
+
+DISPATCHER_TEST(GIVEN_Dispatcher_WHEN_Callback_Added_And_Invoked_THEN_Callback_Invoked_With_Correct_Values)
+{
+	bool Invoked = false;
+	SpatialGDK::FDispatcher Dispatcher;
+	const TUniquePtr<SpatialGDK::EntityView> View = MakeUnique<SpatialGDK::EntityView>();
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:123]: This doesn't look like this needs to be a raw pointer. Just making and EntityView should do fine.

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:123]: done!

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+constexpr Worker_ComponentId OTHER_COMPONENT_ID = 1001;
+constexpr Worker_EntityId ENTITY_ID = 1;
+constexpr Worker_EntityId OTHER_ENTITY_ID = 2;
+constexpr double COMPONENT_VALUE = 3;
+constexpr double OTHER_COMPONENT_VALUE = 4;
+
+void PopulateViewDeltaWithComponentAdded(SpatialGDK::ViewDelta& Delta, SpatialGDK::EntityView& View, const Worker_EntityId EntityId,
+										 const Worker_ComponentId ComponentId, const double Value)
+{
+	SpatialGDK::EntityComponentOpListBuilder OpListBuilder = SpatialGDK::EntityComponentOpListBuilder{};
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:30]: Nit - don't need the equals or stuff after it. This has a default constructor.
General style in spatial view has been to use universal initialiser stuff only for pod or stuff that looks like pod.

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+} // anonymous namespace
+
+DISPATCHER_TEST(GIVEN_Dispatcher_WHEN_Callback_Added_Then_Invoked_THEN_Callback_Invoked_With_Correct_Values)
+{
+	bool Invoked = false;
+	SpatialGDK::FDispatcher Dispatcher;
+	SpatialGDK::EntityView View;
+	SpatialGDK::ViewDelta Delta;
+
+	const SpatialGDK::FComponentValueCallback Callback = [&Invoked](const SpatialGDK::FEntityComponentChange Change) {
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:48]: Is this meant to const ref instead of just const? Same below

File: SpatialGDK/Source/SpatialGDK/Public/SpatialView/Dispatcher.h [hunk 5]
```diff
@@ -20,15 +20,18 @@ using FComponentValueCallback = TCallbacks<FEntityComponentChange>::CallbackType
 class FDispatcher
 {
 public:
-	explicit FDispatcher(const EntityView* View);
+	explicit FDispatcher();
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialView/Dispatcher.h:23]: nit - probably don't need to make it explicit if there is no argument

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+constexpr Worker_EntityId OTHER_ENTITY_ID = 2;
+constexpr double COMPONENT_VALUE = 3;
+constexpr double OTHER_COMPONENT_VALUE = 4;
+
+void PopulateViewDeltaWithComponentAdded(SpatialGDK::ViewDelta& Delta, SpatialGDK::EntityView& View, const Worker_EntityId EntityId,
+										 const Worker_ComponentId ComponentId, const double Value)
+{
+	SpatialGDK::EntityComponentOpListBuilder OpListBuilder = SpatialGDK::EntityComponentOpListBuilder{};
+	OpListBuilder.AddComponent(EntityId, SpatialGDK::CreateTestComponentData(ComponentId, Value));
+	SetFromOpList(Delta, View, OpListBuilder);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:32]: Ah possible driveby I didn't notice from before. This function is transferring ownership so it should take the builder by value. Then it should be moved from here.

**Review by samiwh — APPROVED** [review]
Left a couple of comments and a driveby that can be pushed in a different PR if easier. Otherwise lgtm

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include <memory>
```
Comment (jessicafalk) [inline SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:3]: do we need this include?

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp (hunk 6 above)
Comment (jessicafalk) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp]: I don't think that's possible, iirc I had to pass it by reference as the copy constructor wasn't available for all fields. but maybe there is another way to do it?

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp (hunk 7 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp:3]: no, removed, thanks!

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp (hunk 6 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp]: moving means it isn't being copied- do you mean the oplistbuilder can't be copied? Because then we could force that it has to be moved in instead.

**Review by jessicafalk — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp (hunk 6 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp]: Made the argument to the function an &&

Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/Tests/SpatialView/DispatcherTest.cpp]: and undone
